Digital Credentials: ignore unknown digital credential types

marcoscaceres opened the pull request:
#### b2e1c1159b0077c428d7f5c9e1db7fb062207639
<pre>
Digital Credentials: ignore unknown digital credential types
<a href="https://rdar.apple.com/166673454">rdar://166673454</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=304158">https://bugs.webkit.org/show_bug.cgi?id=304158</a>

Reviewed by Anne van Kesteren.

WebKit relied on the IdentityCredentialProtocol.idl enum to prevent requests with unknown protocols being processed.
However, this prevented unknown protocols from being ignored gracefully, as required by this spec change:
<a href="https://github.com/w3c-fedid/digital-credentials/pull/372">https://github.com/w3c-fedid/digital-credentials/pull/372</a>

We now gracefully ignore unknown protocols by filtering them out, rather than throwing an error.
We also now show a console warning, so developers are aware of ignored protocols.

Includes upstream web platform test commit:
<a href="https://github.com/web-platform-tests/wpt/commit/2d00123f9d72e0c18fa3110aec3893931b00c83b">https://github.com/web-platform-tests/wpt/commit/2d00123f9d72e0c18fa3110aec3893931b00c83b</a>

Tests: http/tests/digital-credentials/digital-credential-console-messages.https.html
       http/wpt/identity/digital-credential-protocol-filtering.https.html

* LayoutTests/http/tests/digital-credentials/digital-credential-console-messages.https-expected.txt: Added.
* LayoutTests/http/tests/digital-credentials/digital-credential-console-messages.https.html: Added.
* LayoutTests/http/wpt/identity/digital-credential-protocol-filtering.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/digital-credential-protocol-filtering.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/create.tentative.https.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/get.tentative.https.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/support/helper.js:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios-18/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/digital-credentials/get.tentative.https-expected.txt:
* LayoutTests/platform/mac-sequoia/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/visionos/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/Modules/identity/DigitalCredential.cpp:
(WebCore::convertProtocolString):
(WebCore::jsToCredentialRequest):
(WebCore::DigitalCredential::convertObjectsToDigitalPresentationRequests):
* Source/WebCore/Modules/identity/DigitalCredentialRequest.h:
* Source/WebCore/Modules/identity/DigitalCredentialRequest.idl:

Canonical link: <a href="https://commits.webkit.org/305257@main">https://commits.webkit.org/305257@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22fef8b480ea1e2e09e2ebd733253a63165e4589

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137947 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10311 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49309 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/146014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/90922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/99f3e227-7810-487d-817a-bf80e93ac191) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139819 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11013 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10453 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/146014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/90922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2534c4a4-487c-4c98-aa7e-4f361e86e26c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140892 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8207 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/146014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1b1d83ea-2443-4c39-b689-e6027ce79871) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7821 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5571 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6296 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117216 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41825 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148724 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9994 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42384 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/113889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8428 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114219 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29011 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7757 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119911 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/64718 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10040 "Built successfully") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9770 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/73608 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9981 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9832 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/Modules/credentialmanagement/CredentialsContainer.cpp:
```diff
@@ -55,10 +55,9 @@ void CredentialsContainer::get(CredentialRequestOptions&& options, CredentialPro
 #if ENABLE(WEB_AUTHN)
     if (options.digital) {
-        DigitalCredential::discoverFromExternalSource(*document(), WTF::move(promise), WTF::move(options));
+        DigitalCredential::discoverFromExternalSource(*m_document, WTF::move(promise), WTF::move(options));
```
This seems rather unrelated and not really consistent either? Still calling `document()` below for instance.

marcoscaceres replied:
Yeah... I was going to ping you about this... the `document.addConsoleMessage()` wants a "non-const" document, which is why I ended up changing the signatures. 

I'll see if there's some way I can get a non const `document` to somehow call `document.addConsoleMessage()` where it needs to happen.

marcoscaceres replied:
Found a better solution...

annevk commented on Source/WebCore/Modules/identity/DigitalCredential.cpp:
```diff
@@ -72,43 +81,53 @@ static ExceptionOr<UnvalidatedDigitalCredentialRequest> jsToCredentialRequest(co
         if (result.hasException(scope)) [[unlikely]]
             return Exception { ExceptionCode::ExistingExceptionError };
-        return DigitalCredentialRequestTypes { WTF::InPlaceType<MobileDocumentRequest>, result.releaseReturnValue() };
+        return std::optional<UnvalidatedDigitalCredentialRequest> { DigitalCredentialRequestTypes { WTF::InPlaceType<MobileDocumentRequest>, result.releaseReturnValue() } };
```
Do we really need all this wrapper boilerplate?

marcoscaceres replied:
Managed to minimize some of the boilerplate. But yes, need some because `ExceptionOr` expects it to be fairly explicit.

annevk commented on LayoutTests/platform/ios/imported/w3c/web-platform-tests/digital-credentials/get.tentative.https-expected.txt:
```diff
@@ -3,16 +3,16 @@
-PASS navigator.credentials.get() API rejects if there are no credential request for same-origin iframe.
+FAIL navigator.credentials.get() API rejects if there are no credential request for same-origin iframe. promise_rejects_js: function "function() { throw e; }" threw object "NotAllowedError: The document is not focused." ("NotAllowedError") expected instance of function "function TypeError() {
+    [native code]
+}" ("TypeError")
```
Seems like the test should be fixed?

marcoscaceres replied:
I'll take a look, but I'm pretty sure this is https://bugs.webkit.org/show_bug.cgi?id=278553 😢

marcoscaceres replied:
Yeah, the test is doing the right thing... it's calling focus()... just WebKit is not focusing it :/

marcoscaceres committed:
Digital Credentials: ignore unknown digital credential types
rdar://166673454
https://bugs.webkit.org/show_bug.cgi?id=304158

Reviewed by Anne van Kesteren.

WebKit relied on the IdentityCredentialProtocol.idl enum to prevent requests with unknown protocols being processed.
However, this prevented unknown protocols from being ignored gracefully, as required by this spec change:
https://github.com/w3c-fedid/digital-credentials/pull/372

We now gracefully ignore unknown protocols by filtering them out, rather than throwing an error.
We also now show a console warning, so developers are aware of ignored protocols.

Includes upstream web platform test commit:
https://github.com/web-platform-tests/wpt/commit/2d00123f9d72e0c18fa3110aec3893931b00c83b

Tests: http/tests/digital-credentials/digital-credential-console-messages.https.html
       http/wpt/identity/digital-credential-protocol-filtering.https.html

* LayoutTests/http/tests/digital-credentials/digital-credential-console-messages.https-expected.txt: Added.
* LayoutTests/http/tests/digital-credentials/digital-credential-console-messages.https.html: Added.
* LayoutTests/http/wpt/identity/digital-credential-protocol-filtering.https-expected.txt: Added.
* LayoutTests/http/wpt/identity/digital-credential-protocol-filtering.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/create.tentative.https.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/get.tentative.https.html:
* LayoutTests/imported/w3c/web-platform-tests/digital-credentials/support/helper.js:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios-18/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/digital-credentials/get.tentative.https-expected.txt:
* LayoutTests/platform/mac-sequoia/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/visionos/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/Modules/identity/DigitalCredential.cpp:
(WebCore::convertProtocolString):
(WebCore::jsToCredentialRequest):
(WebCore::DigitalCredential::convertObjectsToDigitalPresentationRequests):
* Source/WebCore/Modules/identity/DigitalCredentialRequest.h:
* Source/WebCore/Modules/identity/DigitalCredentialRequest.idl:

Canonical link: https://commits.webkit.org/305257@main
